### DISCUSSION TITLE
[Support] v3 如何在运行时动态调整日志等级

### DISCUSSION THREAD

**Discussion by mzzya** [midwayjs/midway#4144] [Q&A]
### Describe the problem（描述问题）

```ts
export const midwayLogger: ServiceFactoryConfigOption<MidwayLoggerOptions> = {
    default: {
        // level: 'info',//默认info
        transports: {
            console: {},
            file: false,
            error: {},
            json: {
                dir: 'log/',
                fileLogName: 'app.log',
                createSymlink: true,
                zippedArchive: true,
                maxSize: '100m',
                maxFiles: '5',
                format: jsonFormat,
            },
        }
    },
    clients: {
        coreLogger: {
            transports: {
                json: {
                    fileLogName: 'core.log',
                }
            }
        },
        appLogger: {
            transports: {
                json: {
                    fileLogName: 'app.log',
                }
            }
        }
    }
}


export class MainConfiguration implements ILifeCycle {
  @App('koa')
  app: koa.Application;

  @Logger()
  logger: MidwayLogger;

  @Inject()
  redisServiceFactory: RedisServiceFactory;

  /**
  * 在应用配置加载后执行
  */
  async onConfigLoad(container: IMidwayContainer, app: IMidwayApplication) {

    const nacosOptions: NacosOptions = this.app.getConfig("mse")

    const nacosClient = new NacosConfigClient({
      serverAddr: nacosOptions.serverAddr,
      namespace: nacosOptions.namespace,
      accessKey: nacosOptions.accessKey,
      secretKey: nacosOptions.secretKey,
      requestTimeout: 6000,
    });

    const configYaml = await nacosClient.getConfig(nacosOptions.dataId, nacosOptions.group)
    const config: any = YAML.parse(configYaml);

    await this.redisServiceFactory.createInstance({ ...config.redis, db: config.redis.database }, "default");
    this.logger.info("项目配置加载完成");
    nacosClient.subscribe({ dataId: nacosOptions.dataId, group: nacosOptions.group }, async (configYamlStr: any) => {
      // console.log("subscribe", dataStr);
      try {
        const config: any = YAML.parse(configYamlStr);
        this.logger.level = 'debug';

        container.registerObject('logger', this.logger);
        this.logger.debug("哈哈哈")

        this.app.addConfigObject({
          //将MSE
          config: config,
        })
        this.logger.debug("配置变更，加载新的应用配置", config)
      } catch (e) {
        console.log("配置转换成JSON异常", e);
      }
    })
  }
}
```

想在nacos中配置日志等级，出现问题时调成debug级别，不知道该怎么处理

### Midway Versions（Midway 版本）

npm http fetch GET 200 https://registry.npmmirror.com/midway-version 135ms (cache revalidated)
>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(3.18.0)
✓ @midwayjs/axios(not installed)
✓ @midwayjs/bootstrap(3.18.2)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/busboy(not installed)
✓ @midwayjs/cache-manager(not installed)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.18.0)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(not installed)
✓ @midwayjs/cross-domain(not installed)
✓ @midwayjs/decorator(not installed)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✓ @midwayjs/i18n(3.18.2)
✓ @midwayjs/info(3.18.2)
✓ @midwayjs/jwt(not installed)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.18.2)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(not installed)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(3.18.2)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(3.18.2)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/tags(not installed)
✓ @midwayjs/tenant(not installed)
✓ @midwayjs/typegoose(not installed)
✓ @midwayjs/typeorm(not installed)
✓ @midwayjs/upload(not installed)
✓ @midwayjs/validate(3.18.2)
✓ @midwayjs/version(not installed)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.18.2)
✓ @midwayjs/ws(not installed)
**********************************************************************
>> Check complete, all versions are healthy.
**********************************************************************

**Comment by mzzya** [comment]
```ts
nacosClient.subscribe({ dataId: nacosOptions.dataId, group: nacosOptions.group }, async (configYamlStr: any) => {
      try {
        const config: any = YAML.parse(configYamlStr);
        const level = config?.logger?.level as LoggerLevel || 'info';

        this.logger.level = level;
        this.logger.get("json").level = level;
        this.logger.get("console").level = level;
        this.logger.get("error").level = level;

        this.app.addConfigObject({
          //将MSE
          config: config,
        })
        this.logger.info("nacos配置加载完成", level == 'debug' ? config : '');
      } catch (e) {
        this.logger.error("配置转换成JSON异常", e, configYamlStr)
      }
    })
```
试出来这样可行，是不是正确用法？

**Comment by czy88840616** [comment]
看着怪怪的，按理日志这个东西就不应该动态改了，毕竟改了之后相当于是全局生效了。

**Reply by mzzya** [reply]
容器化部署的，日志都推送到日志中心去了，线上出问题需要开debug日志看看详细日志。开发和运维觉得重启或者改文件都麻烦，想要java spring那种改下nacos日志logging.level参数就生效的效果。

**Comment by geekberryonekey** [comment]
@czy88840616 类似的问题(`@midwayjs/logger`)
在 onConfigLoad 中返回并不能将默认的日志等级从 info 改为 debug.
onReady 时读取的 config.midwayLogger 是onConfigLoad返回的预期值, 但 this.logger.debug 并不能生效
```
  "midwayLogger": {
    "default": {
      "level": "debug",
      "transports": {
        "console": {
          "autoColors": true,
          "level": "debug"
        }
      }
    }
  }
```

**Reply by czy88840616** [reply]
启动之后改日志的配置没什么用的，直接改某个 logger.level 就好。https://github.com/midwayjs/logger/blob/main/test/index.test.ts#L608

**Comment by czy88840616** [comment]
`onConfigLoad` 里返回/合并 `midwayLogger` 后，不建议理解成“会重新创建 logger”。在 v3 里 loggerFactory / 部分默认 logger 可能已经创建好了，后续只是配置对象变了，已经拿到的 `this.logger`、`appLogger`、`coreLogger` 以及它们下面的 transport 不会自动按新配置重建，所以会出现 `onReady` 读到 config 是 `debug`，但 `this.logger.debug()` 仍然不输出的情况。

如果是启动时就要打开 debug，建议直接在静态配置里同时设置 logger client 和 transport 的 level，例如：

```ts
export const midwayLogger = {
  default: {
    level: 'debug',
    transports: {
      console: {
        level: 'debug',
        autoColors: true,
      },
    },
  },
  clients: {
    appLogger: {
      level: 'debug',
      transports: {
        console: {
          level: 'debug',
        },
      },
    },
    coreLogger: {
      level: 'debug',
      transports: {
        console: {
          level: 'debug',
        },
      },
    },
  },
};
```

如果是运行时从 nacos 动态调整，只改 `config.midwayLogger` 不够，需要改已经存在的 logger 实例和 transport 实例：

```ts
const level = config?.logger?.level || 'info';
const appLogger = this.app.getLogger('appLogger');

appLogger.level = level;
for (const name of ['console', 'file', 'error', 'json']) {
  const transport = appLogger.get?.(name);
  if (transport) {
    transport.level = level;
  }
}
```

如果你用的是 `@Logger('coreLogger')`，也要对 `coreLogger` 做同样处理。这里的关键点是：`logger.level` 和每个 transport 的 `level` 都可能拦截日志；只改其中一个不一定能让 `debug` 输出。

不过动态调日志级别会全局影响当前进程，线上使用时建议限制入口，比如只允许临时打开一段时间，或者只调整某个自定义 logger，避免把 core/app 全局日志长期切到 debug。
